src/main/webapp/app/exercises/shared/plagiarism/plagiarism-split-view/text-submission-viewer/text-submission-viewer.component.ts

```diff
@@ -201,47 +201,86 @@ export class TextSubmissionViewerComponent implements OnChanges {
         const rows = fileContent.split('\n');
-        const matches = this.getMatchesForCurrentFile();
-        const offsets = new Array(rows.length).fill(0);
+        let result = '';
+
+        for (let i = 0; i < matches[0].from.line - 1; i++) {
+            if (rows[i]) {
+                result += escape(rows[i]) + '\n';
+            }
+        }
+        result += escape(rows[matches[0].from.line - 1].slice(0, matches[0].from.column - 1));
 
-        matches.forEach((match) => {
+        for (let i = 0; i < matches.length; i++) {
+            const match = matches[i];
             if (!match.from) {
                 captureException(new Error('"from" is not defined in insertMatchTokens'));
-                return;
+                continue;
             }
             if (!match.to) {
                 captureException(new Error('"to" is not defined in insertMatchTokens'));
-                return;
+                continue;
             }
 
             const idxLineFrom = match.from.line - 1;
             const idxLineTo = match.to.line - 1;
 
-            const idxColumnFrom = match.from.column - 1 + offsets[idxLineFrom];
+            const idxColumnFrom = match.from.column - 1;
+            const idxColumnTo = match.to.column + match.to.length - 1;
 
-            if (rows[idxLineFrom]) {
-                rows[idxLineFrom] = this.insertToken(rows[idxLineFrom], this.tokenStart, idxColumnFrom);
-                offsets[idxLineFrom] += this.tokenStart.length;
+            if (!rows[idxLineFrom]) {
+                continue;
             }
+            result += this.tokenStart;
 
-            const idxColumnTo = match.to.column + match.to.length - 1 + offsets[idxLineTo];
+            if (idxLineFrom === idxLineTo) {
+                result += escape(rows[idxLineFrom].slice(idxColumnFrom, idxColumnTo)) + this.tokenEnd;
+            } else {
+                result += escape(rows[idxLineFrom].slice(idxColumnFrom));
+                for (let j = idxLineFrom + 1; j < idxLineTo; j++) {
+                    if (rows[j]) {
+                        result += '\n' + escape(rows[j]);
+                    }
+                }
+                result += escape(rows[idxLineTo].slice(0, idxColumnTo)) + this.tokenEnd;
+            }
 
-            if (rows[idxLineTo]) {
-                rows[idxLineTo] = this.insertToken(rows[idxLineTo], this.tokenEnd, idxColumnTo);
-                offsets[idxLineTo] += this.tokenEnd.length;
+            // escape everything up until the next match (or the end of the string if there is no more match)
+            if (i === matches.length - 1) {
+                result += escape(rows[idxLineTo].slice(idxColumnTo));
+                for (let j = idxLineTo + 1; j < rows.length; j++) {
+                    if (rows[j]) {
+                        result += '\n' + escape(rows[j]);
+                    }
+                }
+            } else if (matches[i + 1].from.line === match.to.line) {
```

Comment:
What if the next match is an invalid one (with ``from`` missing)? You check this in the beginning of the match loop, but you access it before the next match is checked by these safeguards here.
*If you change the code to filter out invalid matches above, ignore this comment*